### PR TITLE
srp: rebuild library

### PR DESCRIPTION
**Pull request by jbis9051**
This PR is a complete rewrite of the SRP library. It includes many improvements over the old library at the expense of backwards compatibility. 

Improvements include:
- Improved file and code organization
- Access to individual SRP computations 
- Consistent sever and client API
- Simpler API
- Improved documentation with tests in documentation
- New tests for compatibility with the RFC
- Bumps dependencies
- Timing safe verification comparisons
- Modernized error handling 
- And much more...

### CODE REVIEW

**Comment by tarcieri** [comment]
Interesting. Let me know when this is ready for review. I can leave some initial notes.

File: srp/Cargo.toml [hunk 1]
```diff
@@ -15,11 +15,14 @@ num-bigint = "0.4"
 generic-array = "0.14"
 digest = "0.9"
 lazy_static = "1.2"
+subtle = "2.4"
 
 [dev-dependencies]
 rand = "0.6"
 sha2 = "0.9"
 sha-1 = "0.9"
+hex = "0.4"
```
Comment (tarcieri) [inline srp/Cargo.toml:24]: We generally use the `hex-literal` crate for this purpose.

File: srp/Cargo.toml [hunk 2]
```diff
@@ -15,11 +15,14 @@ num-bigint = "0.4"
...
+num-traits = "0.2"
 
 [badges]
 travis-ci = { repository = "RustCrypto/PAKEs" }
```
Comment (tarcieri) [inline srp/Cargo.toml:28]: You can get rid of this. We should use GitHub Actions instead.

File: srp/tests/rfc5054.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        v.to_bytes_be(),
+        hex::decode(strip_white(
+            "        
+         7E273DE8 696FFC4F 4E337D05 B4B375BE B0DDE156 9E8FA00A 9886D812
+         9BADA1F1 822223CA 1A605B53 0E379BA4 729FDC59 F105B478 7E5186F5
+         C671085A 1447B52A 48CF1970 B4FB6F84 00BBF4CE BFBB1681 52E08AB5
+         EA53D15C 1AFF87B2 B9DA6E04 E058AD51 CC72BFC9 033B564E 26480D78
+         E955A5E2 9E7AB245 DB2BE315 E2099AFB
+         "
+        ))
```
Comment (tarcieri) [inline srp/tests/rfc5054.rs:49]: With the `hex-literal` crate this becomes:

```suggestion
        hex!(
            "        
         7E273DE8 696FFC4F 4E337D05 B4B375BE B0DDE156 9E8FA00A 9886D812
         9BADA1F1 822223CA 1A605B53 0E379BA4 729FDC59 F105B478 7E5186F5
         C671085A 1447B52A 48CF1970 B4FB6F84 00BBF4CE BFBB1681 52E08AB5
         EA53D15C 1AFF87B2 B9DA6E04 E058AD51 CC72BFC9 033B564E 26480D78
         E955A5E2 9E7AB245 DB2BE315 E2099AFB
         "
        )
```

It automatically handles stripping whitespace.

**Comment by newpavlov** [comment]
@tarcieri 
Would it be possible to use `crypto-bigint` for implementing SRP or is not not sufficiently advanced yet?

File: srp/Cargo.toml (hunk 2 above)
Comment (newpavlov) [reply srp/Cargo.toml:28]: IIRC badges in `Cargo.toml` got deprecated, so we can simply remove it.

**Comment by tarcieri** [comment]
@newpavlov unfortunately not. The core functionality this needs is modular exponentiation (modpow). We attempted to add that to `crypto-bigint` but never got it working properly.

I might take a crack at it again soon for the purposes of expanding the functionality of `ScalarCore` in the `elliptic-curve` crate. The core algorithm is rather simple: use the existing multiply to produce a wide output, then perform a Montgomery reduction, which requires precomputing *-(m^{-1} mod m) mod m*, but otherwise provides a relatively simple constant-time formula.

**Comment by jbis9051** [comment]
`crypto-bigint` looks really interesting!

**Comment by jbis9051** [comment]
Looks like we will have to change minimum supported rust version

**Comment by tarcieri** [comment]
That's no problem. I might suggest bumping `edition = "2021"` and setting `rust-version = "1.56"` in Cargo.toml.

**Comment by jbis9051** [comment]
PR is complete. I will add some more features in a future.

**Comment by jbis9051** [comment]
I'm not sure how to handle the failing GH workflows. It seems they aren't scoped to the proper sub-crate so SRP min rust version is interfering with spake2.

**Comment by tarcieri** [comment]
Aah, this is the joy of having one 2021 edition crate in a workspace. It either needs to be split into its own workspace, or we should bump all of the crates to MSRV 1.56.

I can take care of that, if you'd like.

Clippy, on the other hand, is global to the whole workspace and needs an MSRV of the highest MSRV of any crate in the repo.

**Comment by jbis9051** [comment]
Could we do something like

```spake2.yml
- run: cd spake2 && cargo test --release
```

?

>  It either needs to be split into its own workspace, or we should bump all of the crates to MSRV 1.56.

If not, I'd say bump.

> I can take care of that, if you'd like.

Sure. Whatever you think this best.

**Comment by tarcieri** [comment]
It already does that:

https://github.com/RustCrypto/PAKEs/blob/master/.github/workflows/spake2.yml#L11-L13

...however you're running into things which are inherently global to a workspace, regardless of the CI config

**Comment by newpavlov** [comment]
Yeah, I think the easiest solution would be to simply bump `spake2`'s MSRV to 1.56.

**Comment by tarcieri** [comment]
I bumped both `spake2` and `srp` to Rust 2021 edition and did various other minor documentation and clippy fixups in #80.

@jbis9051 if you rebase, you should be good to go now

**Comment by jbis9051** [comment]
\o/

**Comment by tarcieri** [comment]
Success!

I’ll try to look through this in the coming weeks, but expect some delays there due to the holidays

**Comment by jbis9051** [comment]
Sounds good and happy holidays!

**Comment by jbis9051** [comment]
Hi. Any update on this?

**Comment by tarcieri** [comment]
Will try to review it this weekend

**Review by tarcieri — APPROVED** [review]
This looks like an improvement to me.

I say this with the caveat that I'm not particularly familiar with SRP, so I can't say my review was particularly careful or detailed, nor do I have any specific feedback beyond what's already been covered earlier.

I'll give @newpavlov a bit of time to review this, or if he's too busy I'll merge it next weekend.

**Comment by tarcieri** [comment]
@jbis9051 thank you! I can go ahead and cut a release if you'd like

**Comment by jbis9051** [comment]
@tarcieri  Thank you! Sure that would be great.

P.S. This PR (and the last one) closed the following issues: https://github.com/RustCrypto/PAKEs/issues/17, https://github.com/RustCrypto/PAKEs/issues/18

**Pull request merged**